{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "图片预处理及增强"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "from keras.applications import xception\n",
    "from keras.preprocessing.image import ImageDataGenerator\n",
    "\n",
    "def img_preprocessing(val_path,tra_path):\n",
    "    img_height, img_width=(299,299)\n",
    "    xception_train_datagen = ImageDataGenerator(zoom_range=[0.8,1.2],\n",
    "                                           rotation_range=10,\n",
    "                                           width_shift_range=0.2,\n",
    "                                           height_shift_range=0.2,\n",
    "                                           preprocessing_function=xception.preprocess_input)\n",
    "    xception_validation_datagen = ImageDataGenerator(preprocessing_function=xception.preprocess_input)\n",
    "\n",
    "    print(\"\")\n",
    "    print(\"xception train/validation dataset loading\")\n",
    "    print(\"------------------------------------\")\n",
    "    print(\"\")\n",
    "    xception_train_generator = xception_train_datagen.flow_from_directory(\n",
    "            tra_path,\n",
    "            target_size=( img_height, img_width),\n",
    "            batch_size=32,\n",
    "            class_mode='categorical')\n",
    "    xception_validation_generator = xception_validation_datagen.flow_from_directory(\n",
    "            val_path,\n",
    "            target_size=( img_height, img_width),\n",
    "            batch_size=32,\n",
    "            class_mode='categorical',\n",
    "            shuffle=False)\n",
    "    return xception_train_generator,xception_validation_generator"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "模型搭建"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from keras.applications.xception import Xception\n",
    "from keras.preprocessing import image\n",
    "from keras.models import Model\n",
    "from keras.layers import GlobalAveragePooling2D,Dense,Dropout\n",
    "from keras import backend as K\n",
    "from keras import optimizers\n",
    "from keras import regularizers\n",
    "from keras.callbacks import ModelCheckpoint\n",
    "from keras.callbacks import EarlyStopping\n",
    "\n",
    "def xception_train(path,xception_train_generator,xception_validation_generator,lr,dec):\n",
    "    xception_base_model = Xception(weights='imagenet', include_top=False)\n",
    "    x_xception = xception_base_model.output\n",
    "    x_xception = GlobalAveragePooling2D()(x_xception)\n",
    "    x_xception = Dropout(0.4)(x_xception)\n",
    "    predictions = Dense(10, activation='softmax')(x_xception)\n",
    "    xception_model = Model(inputs=xception_base_model.input, outputs=predictions)\n",
    "    for layer in xception_base_model.layers[:249]:\n",
    "        layer.trainable = False\n",
    "    for layer in xception_base_model.layers[249:]:\n",
    "        layer.trainable = True    \n",
    "    sgd = optimizers.SGD(lr=lr,decay=dec, momentum=0.9, nesterov=True)\n",
    "    xception_model.compile(optimizer=sgd, loss='categorical_crossentropy',metrics=['categorical_accuracy'])\n",
    "    CHECKPOINT_MODEL_SAVE_PATH = os.path.join(path,\"xception_best_model_{epoch:04d}_{val_loss:.4f}.hdf5\")\n",
    "    model_check_point =ModelCheckpoint(CHECKPOINT_MODEL_SAVE_PATH,monitor='val_loss',verbose=1,save_best_only=True,save_weights_only=False,period=1) \n",
    "    model_early_stop=EarlyStopping(monitor='val_loss', patience=5, verbose=1, mode='min')\n",
    "    print(\"\")\n",
    "    print(\"xception training\")\n",
    "    print(\"------------------------------------\")\n",
    "    print(\"\")\n",
    "    xception_model.fit_generator(\n",
    "                            xception_train_generator,\n",
    "                            epochs=30,\n",
    "                            validation_data=xception_validation_generator,\n",
    "                            callbacks=[model_check_point,model_early_stop]\n",
    "                            )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "模型训练"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "t4 t4\\my_validation t4\\my_train 0.001 0.0001\n",
      "------------------------------------\n",
      "\n",
      "xception train/validation dataset loading\n",
      "------------------------------------\n",
      "\n",
      "Found 20408 images belonging to 10 classes.\n",
      "Found 2016 images belonging to 10 classes.\n",
      "\n",
      "xception training\n",
      "------------------------------------\n",
      "\n",
      "Epoch 1/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 2.1868 - categorical_accuracy: 0.2178\n",
      "Epoch 00001: val_loss improved from inf to 2.23542, saving model to t4\\xception_best_model_0001_2.2354.hdf5\n",
      "638/638 [==============================] - 507s 795ms/step - loss: 2.1867 - categorical_accuracy: 0.2179 - val_loss: 2.2354 - val_categorical_accuracy: 0.1677\n",
      "Epoch 2/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.9580 - categorical_accuracy: 0.3834\n",
      "Epoch 00002: val_loss did not improve\n",
      "638/638 [==============================] - 501s 785ms/step - loss: 1.9578 - categorical_accuracy: 0.3836 - val_loss: 2.2565 - val_categorical_accuracy: 0.2103\n",
      "Epoch 3/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.8133 - categorical_accuracy: 0.4456\n",
      "Epoch 00003: val_loss did not improve\n",
      "638/638 [==============================] - 501s 786ms/step - loss: 1.8132 - categorical_accuracy: 0.4456 - val_loss: 2.2997 - val_categorical_accuracy: 0.2555\n",
      "Epoch 4/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.7097 - categorical_accuracy: 0.4898\n",
      "Epoch 00004: val_loss did not improve\n",
      "638/638 [==============================] - 501s 786ms/step - loss: 1.7097 - categorical_accuracy: 0.4897 - val_loss: 2.3637 - val_categorical_accuracy: 0.2773\n",
      "Epoch 5/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.6290 - categorical_accuracy: 0.5110\n",
      "Epoch 00005: val_loss did not improve\n",
      "638/638 [==============================] - 501s 786ms/step - loss: 1.6291 - categorical_accuracy: 0.5109 - val_loss: 2.3598 - val_categorical_accuracy: 0.2882\n",
      "Epoch 6/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.5770 - categorical_accuracy: 0.5273\n",
      "Epoch 00006: val_loss did not improve\n",
      "638/638 [==============================] - 501s 786ms/step - loss: 1.5769 - categorical_accuracy: 0.5273 - val_loss: 2.3984 - val_categorical_accuracy: 0.2788\n",
      "Epoch 00006: early stopping\n",
      "t4 t4\\my_validation t4\\my_train 0.001 1e-05\n",
      "------------------------------------\n",
      "\n",
      "xception train/validation dataset loading\n",
      "------------------------------------\n",
      "\n",
      "Found 20408 images belonging to 10 classes.\n",
      "Found 2016 images belonging to 10 classes.\n",
      "\n",
      "xception training\n",
      "------------------------------------\n",
      "\n",
      "Epoch 1/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 2.1890 - categorical_accuracy: 0.2152\n",
      "Epoch 00001: val_loss improved from inf to 2.26258, saving model to t4\\xception_best_model_0001_2.2626.hdf5\n",
      "638/638 [==============================] - 507s 795ms/step - loss: 2.1886 - categorical_accuracy: 0.2155 - val_loss: 2.2626 - val_categorical_accuracy: 0.1925\n",
      "Epoch 2/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.9567 - categorical_accuracy: 0.3807\n",
      "Epoch 00002: val_loss did not improve\n",
      "638/638 [==============================] - 502s 786ms/step - loss: 1.9565 - categorical_accuracy: 0.3808 - val_loss: 2.2835 - val_categorical_accuracy: 0.1974\n",
      "Epoch 3/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.8023 - categorical_accuracy: 0.4487\n",
      "Epoch 00003: val_loss did not improve\n",
      "638/638 [==============================] - 502s 786ms/step - loss: 1.8022 - categorical_accuracy: 0.4486 - val_loss: 2.3216 - val_categorical_accuracy: 0.1612\n",
      "Epoch 4/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.6854 - categorical_accuracy: 0.4894\n",
      "Epoch 00004: val_loss did not improve\n",
      "638/638 [==============================] - 502s 786ms/step - loss: 1.6855 - categorical_accuracy: 0.4892 - val_loss: 2.3348 - val_categorical_accuracy: 0.1979\n",
      "Epoch 5/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.6074 - categorical_accuracy: 0.5129\n",
      "Epoch 00005: val_loss did not improve\n",
      "638/638 [==============================] - 502s 786ms/step - loss: 1.6070 - categorical_accuracy: 0.5132 - val_loss: 2.3776 - val_categorical_accuracy: 0.1657\n",
      "Epoch 6/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.5332 - categorical_accuracy: 0.5365\n",
      "Epoch 00006: val_loss did not improve\n",
      "638/638 [==============================] - 502s 786ms/step - loss: 1.5334 - categorical_accuracy: 0.5364 - val_loss: 2.4428 - val_categorical_accuracy: 0.2207\n",
      "Epoch 00006: early stopping\n",
      "t4 t4\\my_validation t4\\my_train 0.001 1e-06\n",
      "------------------------------------\n",
      "\n",
      "xception train/validation dataset loading\n",
      "------------------------------------\n",
      "\n",
      "Found 20408 images belonging to 10 classes.\n",
      "Found 2016 images belonging to 10 classes.\n",
      "\n",
      "xception training\n",
      "------------------------------------\n",
      "\n",
      "Epoch 1/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 2.1969 - categorical_accuracy: 0.2105\n",
      "Epoch 00001: val_loss improved from inf to 2.28006, saving model to t4\\xception_best_model_0001_2.2801.hdf5\n",
      "638/638 [==============================] - 510s 800ms/step - loss: 2.1966 - categorical_accuracy: 0.2108 - val_loss: 2.2801 - val_categorical_accuracy: 0.1726\n",
      "Epoch 2/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.9597 - categorical_accuracy: 0.3774\n",
      "Epoch 00002: val_loss did not improve\n",
      "638/638 [==============================] - 502s 786ms/step - loss: 1.9596 - categorical_accuracy: 0.3774 - val_loss: 2.2806 - val_categorical_accuracy: 0.2485\n",
      "Epoch 3/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.7978 - categorical_accuracy: 0.4536\n",
      "Epoch 00003: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.7977 - categorical_accuracy: 0.4536 - val_loss: 2.3773 - val_categorical_accuracy: 0.2004\n",
      "Epoch 4/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.6880 - categorical_accuracy: 0.4897\n",
      "Epoch 00004: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.6881 - categorical_accuracy: 0.4895 - val_loss: 2.3867 - val_categorical_accuracy: 0.2500\n",
      "Epoch 5/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.5969 - categorical_accuracy: 0.5204\n",
      "Epoch 00005: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.5971 - categorical_accuracy: 0.5202 - val_loss: 2.4145 - val_categorical_accuracy: 0.2341\n",
      "Epoch 6/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.5375 - categorical_accuracy: 0.5331\n",
      "Epoch 00006: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.5377 - categorical_accuracy: 0.5330 - val_loss: 2.4508 - val_categorical_accuracy: 0.2495\n",
      "Epoch 00006: early stopping\n",
      "t4 t4\\my_validation t4\\my_train 0.01 0.0001\n",
      "------------------------------------\n",
      "\n",
      "xception train/validation dataset loading\n",
      "------------------------------------\n",
      "\n",
      "Found 20408 images belonging to 10 classes.\n",
      "Found 2016 images belonging to 10 classes.\n",
      "\n",
      "xception training\n",
      "------------------------------------\n",
      "\n",
      "Epoch 1/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.6941 - categorical_accuracy: 0.4359\n",
      "Epoch 00001: val_loss improved from inf to 2.55874, saving model to t4\\xception_best_model_0001_2.5587.hdf5\n",
      "638/638 [==============================] - 512s 802ms/step - loss: 1.6938 - categorical_accuracy: 0.4360 - val_loss: 2.5587 - val_categorical_accuracy: 0.2525\n",
      "Epoch 2/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.2942 - categorical_accuracy: 0.5810\n",
      "Epoch 00002: val_loss improved from 2.55874 to 2.49820, saving model to t4\\xception_best_model_0002_2.4982.hdf5\n",
      "638/638 [==============================] - 503s 788ms/step - loss: 1.2937 - categorical_accuracy: 0.5812 - val_loss: 2.4982 - val_categorical_accuracy: 0.2569\n",
      "Epoch 3/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.1887 - categorical_accuracy: 0.6082\n",
      "Epoch 00003: val_loss improved from 2.49820 to 2.37745, saving model to t4\\xception_best_model_0003_2.3774.hdf5\n",
      "638/638 [==============================] - 503s 788ms/step - loss: 1.1888 - categorical_accuracy: 0.6081 - val_loss: 2.3774 - val_categorical_accuracy: 0.2803\n",
      "Epoch 4/30\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "637/638 [============================>.] - ETA: 0s - loss: 1.1407 - categorical_accuracy: 0.6215\n",
      "Epoch 00004: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.1404 - categorical_accuracy: 0.6217 - val_loss: 2.4917 - val_categorical_accuracy: 0.2892\n",
      "Epoch 5/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.1009 - categorical_accuracy: 0.6321\n",
      "Epoch 00005: val_loss did not improve\n",
      "638/638 [==============================] - 502s 786ms/step - loss: 1.1009 - categorical_accuracy: 0.6320 - val_loss: 2.4782 - val_categorical_accuracy: 0.2892\n",
      "Epoch 6/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.0830 - categorical_accuracy: 0.6348\n",
      "Epoch 00006: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.0828 - categorical_accuracy: 0.6349 - val_loss: 2.4677 - val_categorical_accuracy: 0.2971\n",
      "Epoch 7/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.0708 - categorical_accuracy: 0.6374\n",
      "Epoch 00007: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.0711 - categorical_accuracy: 0.6374 - val_loss: 2.4544 - val_categorical_accuracy: 0.2882\n",
      "Epoch 8/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.0557 - categorical_accuracy: 0.6437\n",
      "Epoch 00008: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.0557 - categorical_accuracy: 0.6438 - val_loss: 2.5731 - val_categorical_accuracy: 0.2887\n",
      "Epoch 00008: early stopping\n",
      "t4 t4\\my_validation t4\\my_train 0.01 1e-05\n",
      "------------------------------------\n",
      "\n",
      "xception train/validation dataset loading\n",
      "------------------------------------\n",
      "\n",
      "Found 20408 images belonging to 10 classes.\n",
      "Found 2016 images belonging to 10 classes.\n",
      "\n",
      "xception training\n",
      "------------------------------------\n",
      "\n",
      "Epoch 1/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.6933 - categorical_accuracy: 0.4397\n",
      "Epoch 00001: val_loss improved from inf to 2.47244, saving model to t4\\xception_best_model_0001_2.4724.hdf5\n",
      "638/638 [==============================] - 514s 805ms/step - loss: 1.6925 - categorical_accuracy: 0.4400 - val_loss: 2.4724 - val_categorical_accuracy: 0.2495\n",
      "Epoch 2/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.2883 - categorical_accuracy: 0.5806\n",
      "Epoch 00002: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.2879 - categorical_accuracy: 0.5807 - val_loss: 2.5800 - val_categorical_accuracy: 0.2976\n",
      "Epoch 3/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.1882 - categorical_accuracy: 0.6085\n",
      "Epoch 00003: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.1892 - categorical_accuracy: 0.6082 - val_loss: 2.7154 - val_categorical_accuracy: 0.2857\n",
      "Epoch 4/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.1202 - categorical_accuracy: 0.6241\n",
      "Epoch 00004: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.1202 - categorical_accuracy: 0.6241 - val_loss: 2.5446 - val_categorical_accuracy: 0.2971\n",
      "Epoch 5/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.0956 - categorical_accuracy: 0.6321\n",
      "Epoch 00005: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.0950 - categorical_accuracy: 0.6324 - val_loss: 2.7752 - val_categorical_accuracy: 0.2624\n",
      "Epoch 6/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.0730 - categorical_accuracy: 0.6371\n",
      "Epoch 00006: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.0732 - categorical_accuracy: 0.6370 - val_loss: 2.7904 - val_categorical_accuracy: 0.2545\n",
      "Epoch 00006: early stopping\n",
      "t4 t4\\my_validation t4\\my_train 0.01 1e-06\n",
      "------------------------------------\n",
      "\n",
      "xception train/validation dataset loading\n",
      "------------------------------------\n",
      "\n",
      "Found 20408 images belonging to 10 classes.\n",
      "Found 2016 images belonging to 10 classes.\n",
      "\n",
      "xception training\n",
      "------------------------------------\n",
      "\n",
      "Epoch 1/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.6778 - categorical_accuracy: 0.4452\n",
      "Epoch 00001: val_loss improved from inf to 2.53300, saving model to t4\\xception_best_model_0001_2.5330.hdf5\n",
      "638/638 [==============================] - 516s 809ms/step - loss: 1.6769 - categorical_accuracy: 0.4455 - val_loss: 2.5330 - val_categorical_accuracy: 0.2391\n",
      "Epoch 2/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.2996 - categorical_accuracy: 0.5757\n",
      "Epoch 00002: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.2994 - categorical_accuracy: 0.5760 - val_loss: 2.6773 - val_categorical_accuracy: 0.2525\n",
      "Epoch 3/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.1784 - categorical_accuracy: 0.6103\n",
      "Epoch 00003: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.1780 - categorical_accuracy: 0.6106 - val_loss: 2.7026 - val_categorical_accuracy: 0.2450\n",
      "Epoch 4/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.1206 - categorical_accuracy: 0.6234\n",
      "Epoch 00004: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.1204 - categorical_accuracy: 0.6234 - val_loss: 2.7625 - val_categorical_accuracy: 0.2356\n",
      "Epoch 5/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.1015 - categorical_accuracy: 0.6339\n",
      "Epoch 00005: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.1015 - categorical_accuracy: 0.6339 - val_loss: 2.6838 - val_categorical_accuracy: 0.2594\n",
      "Epoch 6/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.0839 - categorical_accuracy: 0.6305\n",
      "Epoch 00006: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.0841 - categorical_accuracy: 0.6304 - val_loss: 2.7656 - val_categorical_accuracy: 0.2480\n",
      "Epoch 00006: early stopping\n",
      "t4 t4\\my_validation t4\\my_train 0.03 0.0001\n",
      "------------------------------------\n",
      "\n",
      "xception train/validation dataset loading\n",
      "------------------------------------\n",
      "\n",
      "Found 20408 images belonging to 10 classes.\n",
      "Found 2016 images belonging to 10 classes.\n",
      "\n",
      "xception training\n",
      "------------------------------------\n",
      "\n",
      "Epoch 1/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.4616 - categorical_accuracy: 0.4951\n",
      "Epoch 00001: val_loss improved from inf to 2.44104, saving model to t4\\xception_best_model_0001_2.4410.hdf5\n",
      "638/638 [==============================] - 518s 812ms/step - loss: 1.4613 - categorical_accuracy: 0.4953 - val_loss: 2.4410 - val_categorical_accuracy: 0.3363\n",
      "Epoch 2/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.1875 - categorical_accuracy: 0.5966\n",
      "Epoch 00002: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.1874 - categorical_accuracy: 0.5966 - val_loss: 2.5232 - val_categorical_accuracy: 0.3234\n",
      "Epoch 3/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.1292 - categorical_accuracy: 0.6154\n",
      "Epoch 00003: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.1293 - categorical_accuracy: 0.6153 - val_loss: 2.7750 - val_categorical_accuracy: 0.2713\n",
      "Epoch 4/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.1136 - categorical_accuracy: 0.6223\n",
      "Epoch 00004: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.1136 - categorical_accuracy: 0.6222 - val_loss: 2.9657 - val_categorical_accuracy: 0.3026\n",
      "Epoch 5/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.1073 - categorical_accuracy: 0.6220\n",
      "Epoch 00005: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.1073 - categorical_accuracy: 0.6220 - val_loss: 2.8186 - val_categorical_accuracy: 0.2708\n",
      "Epoch 6/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.0881 - categorical_accuracy: 0.6341\n",
      "Epoch 00006: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.0881 - categorical_accuracy: 0.6340 - val_loss: 2.7984 - val_categorical_accuracy: 0.2932\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 00006: early stopping\n",
      "t4 t4\\my_validation t4\\my_train 0.03 1e-05\n",
      "------------------------------------\n",
      "\n",
      "xception train/validation dataset loading\n",
      "------------------------------------\n",
      "\n",
      "Found 20408 images belonging to 10 classes.\n",
      "Found 2016 images belonging to 10 classes.\n",
      "\n",
      "xception training\n",
      "------------------------------------\n",
      "\n",
      "Epoch 1/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.4576 - categorical_accuracy: 0.4967\n",
      "Epoch 00001: val_loss improved from inf to 2.81820, saving model to t4\\xception_best_model_0001_2.8182.hdf5\n",
      "638/638 [==============================] - 520s 816ms/step - loss: 1.4572 - categorical_accuracy: 0.4968 - val_loss: 2.8182 - val_categorical_accuracy: 0.1944\n",
      "Epoch 2/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.1788 - categorical_accuracy: 0.5982\n",
      "Epoch 00002: val_loss improved from 2.81820 to 2.43964, saving model to t4\\xception_best_model_0002_2.4396.hdf5\n",
      "638/638 [==============================] - 503s 788ms/step - loss: 1.1785 - categorical_accuracy: 0.5985 - val_loss: 2.4396 - val_categorical_accuracy: 0.3110\n",
      "Epoch 3/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.1518 - categorical_accuracy: 0.6104\n",
      "Epoch 00003: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.1519 - categorical_accuracy: 0.6103 - val_loss: 2.7135 - val_categorical_accuracy: 0.2688\n",
      "Epoch 4/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.1289 - categorical_accuracy: 0.6155\n",
      "Epoch 00004: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.1289 - categorical_accuracy: 0.6155 - val_loss: 2.9443 - val_categorical_accuracy: 0.2331\n",
      "Epoch 5/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.1267 - categorical_accuracy: 0.6174\n",
      "Epoch 00005: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.1272 - categorical_accuracy: 0.6173 - val_loss: 2.7759 - val_categorical_accuracy: 0.2932\n",
      "Epoch 6/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.1251 - categorical_accuracy: 0.6252\n",
      "Epoch 00006: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.1257 - categorical_accuracy: 0.6250 - val_loss: 3.7958 - val_categorical_accuracy: 0.2153\n",
      "Epoch 7/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.1181 - categorical_accuracy: 0.6234\n",
      "Epoch 00007: val_loss did not improve\n",
      "638/638 [==============================] - 502s 787ms/step - loss: 1.1178 - categorical_accuracy: 0.6233 - val_loss: 3.5744 - val_categorical_accuracy: 0.2455\n",
      "Epoch 00007: early stopping\n",
      "t4 t4\\my_validation t4\\my_train 0.03 1e-06\n",
      "------------------------------------\n",
      "\n",
      "xception train/validation dataset loading\n",
      "------------------------------------\n",
      "\n",
      "Found 20408 images belonging to 10 classes.\n",
      "Found 2016 images belonging to 10 classes.\n",
      "\n",
      "xception training\n",
      "------------------------------------\n",
      "\n",
      "Epoch 1/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.4578 - categorical_accuracy: 0.5040\n",
      "Epoch 00001: val_loss improved from inf to 2.67989, saving model to t4\\xception_best_model_0001_2.6799.hdf5\n",
      "638/638 [==============================] - 523s 819ms/step - loss: 1.4572 - categorical_accuracy: 0.5041 - val_loss: 2.6799 - val_categorical_accuracy: 0.2956\n",
      "Epoch 2/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.1979 - categorical_accuracy: 0.5937\n",
      "Epoch 00002: val_loss improved from 2.67989 to 2.58467, saving model to t4\\xception_best_model_0002_2.5847.hdf5\n",
      "638/638 [==============================] - 503s 788ms/step - loss: 1.1977 - categorical_accuracy: 0.5938 - val_loss: 2.5847 - val_categorical_accuracy: 0.2877\n",
      "Epoch 3/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.1422 - categorical_accuracy: 0.6105\n",
      "Epoch 00003: val_loss did not improve\n",
      "638/638 [==============================] - 503s 788ms/step - loss: 1.1415 - categorical_accuracy: 0.6108 - val_loss: 2.9122 - val_categorical_accuracy: 0.2659\n",
      "Epoch 4/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.1375 - categorical_accuracy: 0.6145\n",
      "Epoch 00004: val_loss improved from 2.58467 to 2.46634, saving model to t4\\xception_best_model_0004_2.4663.hdf5\n",
      "638/638 [==============================] - 503s 788ms/step - loss: 1.1379 - categorical_accuracy: 0.6145 - val_loss: 2.4663 - val_categorical_accuracy: 0.3026\n",
      "Epoch 5/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.1247 - categorical_accuracy: 0.6193\n",
      "Epoch 00005: val_loss did not improve\n",
      "638/638 [==============================] - 502s 788ms/step - loss: 1.1246 - categorical_accuracy: 0.6192 - val_loss: 2.9233 - val_categorical_accuracy: 0.2361\n",
      "Epoch 6/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.1488 - categorical_accuracy: 0.6128\n",
      "Epoch 00006: val_loss did not improve\n",
      "638/638 [==============================] - 503s 789ms/step - loss: 1.1486 - categorical_accuracy: 0.6129 - val_loss: 3.2000 - val_categorical_accuracy: 0.2688\n",
      "Epoch 7/30\n",
      "637/638 [============================>.] - ETA: 0s - loss: 1.1248 - categorical_accuracy: 0.6224\n",
      "Epoch 00007: val_loss did not improve\n",
      "638/638 [==============================] - 502s 788ms/step - loss: 1.1247 - categorical_accuracy: 0.6224 - val_loss: 2.8980 - val_categorical_accuracy: 0.2619\n",
      "Epoch 8/30\n",
      "538/638 [========================>.....] - ETA: 1:10 - loss: 1.1228 - categorical_accuracy: 0.6253"
     ]
    }
   ],
   "source": [
    "train_folder = 'train'\n",
    "test_folder = 'test'\n",
    "my_train_folder='my_train'\n",
    "my_validation_folder = 'my_validation'\n",
    "path_num=5\n",
    "\n",
    "import os\n",
    "import glob\n",
    "\n",
    "for lr in [1e-3,1e-2,3e-2]:\n",
    "    for dec in [1e-4,1e-5,1e-6]:\n",
    "        path_i = os.path.join('t' + str(4))\n",
    "        v_path=os.path.join(path_i,my_validation_folder)\n",
    "        t_path=os.path.join(path_i,my_train_folder)\n",
    "        print(path_i,v_path,t_path,lr,dec)\n",
    "        print(\"------------------------------------\")\n",
    "        xception_train_generator,xception_validation_generator=img_preprocessing(v_path,t_path)\n",
    "        xception_train (path_i,xception_train_generator,xception_validation_generator,lr,dec)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
